lib/dialects/abstract/query-generator/composition.js

```diff
@@ -0,0 +1,153 @@
+'use strict';
+
+const logger = require('../../../utils/logger');
+const { CompositionError } = require('../../../errors');
+
+/*
+ * Self contained value representation
+ *
+ * It can be converted to an escaped sql string or to a binded param. It also
+ * must be self contained, the escaper or binder should be part of the slot. It
+ * is identified for being an instace of this class, otherwise reject (reflects
+ * unsafe handling)
+ *
+ */
+class Slot {
+  constructor(value, field, options) {
+    this.value = value;
+    this.field = field;
```

Comment:
Why `field` is required, what does `options` contain

Reply:
A `slot` doesn't know if it will be binded or stringified. Depending on whether the slot is binded or stringified, the final value for some dataTypes will be different. For instance `N'abc'` if stringified or just `abc` if binded.

The field argument provides the `stringify` and `bindParam` methods. The `options` argument is used to as `field.stringify(value, options)`.

This helps dealing with very complex dataTypes. For instance, take a look at `Array.prototype._stringify` inside `lib/dialects/postgres/data-types.js`. This function if stringified requires to know the database field name, which is passed through options.